### PR TITLE
Moved pod_update to a pod ginkgo test suite. #3941

### PR DESCRIPTION
**Pull request by rrati**

### CODE REVIEW

**Comment by satnam6502** [comment]
@filbranden

**Comment by satnam6502** [comment]
Kicking Travis to restart.

**Comment by filbranden** [comment]
Can we keep the same file name and the same test name "TestPodUpdate" ?

The same filename would make it easier to review the code to see what was kept and what was changed...

The name on Describe() is what we're indexing our tests with in Jenkins and also for the `cmd/e2e -t` parameter, so I'd really prefer if we didn't change test names while converting them...

Thanks!
Filipe

**Comment by rrati** [comment]
I changed the file name because a ginkgo test suite usually isn't just 1 
test and I was creating a file for all the specs that will test pods. 
That's also why I used a different name in the Describe() line.  Is the 
intention not to move to that format?

There's actually 2 tests in the new file instead of just the singular 
one that was in the original file.  I wouldn't say the added test 
(submit and remove) is really adding anything that the update test 
doesn't cover though.  It probably makes sense to just remove that test.

Is it possible to use something else other than the Describe() entry for 
test indexing?  Maybe that's a later step in the conversion?

Rob

**Comment by filbranden** [comment]
@rrati Yeah let's do one step at a time please!

We're barely having the ability to collect these JUnit XML reports in Jenkins now, before we talk about renaming tests I want to make sure we're getting useful data from Jenkins. So please keep the test name in Describe() the same as it is now.

Also, if you want to rename files, do that in a separate step. If they files look enough alike git will detect it's a rename, but if you rename it *and* change a bunch of things in the same commit then it will have a harder time figuring it out...

Baby steps... baby steps.

**Comment by zmerlynn** [comment]
Actually, Jenkins has no trouble collecting multiple suites right now, unless you think there's some issue in driver.go with outputting them? I really don't mind the conversion being to the final resting place.

**Comment by zmerlynn** [comment]
FWIW, Jenkins glommed far more onto the string "Kubernetes e2e Suite", so the "description" string out of `RunSpecs` seems to dictate the intermediate level of its reporting. The "suite" aspect in Gingko's JUnit reporting somehow (we might want to look into this). Regardless, this structure looks fine, and this test actually splits things up into two assertions, which is cool.

LGTM. Kicking Travis.

**Comment by filbranden** [comment]
Please don't merge it as is.

At the very least, the two commits should be squashed together.

I'm also strongly against renaming the file on this same commit. I really want GitHub to detect a file rename and the best way to do it is in its own commit (you could do it before the conversion or after the conversion, that's about the same.)

I'm against the TestPodUpdate -> Pod rename but if you feel strongly about doing the rename then that's fine. But then instead of doing "one off" renames and having old-style and new-style living together for a long time I'd rather have a single rename of everything on its own commit. You could do it now before the conversion if you'd like.

Cheers,
Filipe

**Comment by rrati** [comment]
I've no problems creating a separate PR for changing the name of the file if that's required.  I think a discussion of the end goal and the steps we'd like to take to get there might be in order.  My assumption has been this ginkgo transition would result in a multiple specs in a single file, ie test suites.  In that situation having "TestPodUpdate" as the name of the suite is misleading since the suite will be testing more than just pod update.  I think a quick hangout could help clarify the steps and end goals here.

**Comment by zmerlynn** [comment]
At the very least you might want an initial, separate commit that does the rename. It's possible that if you do the git mv in one commit and the changes in another that they can still sit in the same PR.

I actually have nothing against this PR. I'd like to cleanup the file tracking so history isn't lost (so agree with Filipe there), and it looks like Travis is still red (squash your gofmt piece back in?).

Maybe IRC later?

File: test/e2e/pod.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Pods", func() {
+	var (
+		c   *client.Client
+		err error
```
Comment (roberthbailey) [inline test/e2e/pod.go]: Why is err a test-wide var?

File: test/e2e/pod.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		pod.Name = pod.Name + "-" + randomSuffix()
+		pod.Labels["time"] = value
+		pod.Spec.Containers[0].Ports[0].HostPort = 0
+
+		By("submitting the pod to kubernetes")
+		_, err = podClient.Create(pod)
+		if err != nil {
+			Fail(fmt.Sprintf("Failed to create pod: %v", err))
+		}
+		defer podClient.Delete(pod.Name)
```
Comment (roberthbailey) [inline test/e2e/pod.go]: The should update a pod test below also calls defer GinkgoRecover() inside of this defer. The two specs should be consistent.

File: test/e2e/pod.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	var (
+		c   *client.Client
+		err error
+	)
+
+	BeforeEach(func() {
+		c = loadClientOrDie()
+	})
+
+	It("should submit and remove a pod", func() {
```
Comment (roberthbailey) [inline test/e2e/pod.go]: The first three steps (load, submit to k8s, verify) are exactly the same code between these two tests. It seems like they should be extracted into a common helper so prevent drift (which we already have with the different defer functions to delete the pod).

**Comment by filbranden** [comment]
I like what I see here.

Looks like merging #3995 caused a conflict on this one... sorry about that. Would you like to rebase it and bring those changes in? If you'd like, I can do a manual merge of this branch and resolve the conflicts as I do that.

@roberthbailey Do you still have comments on the code? Are you satisfied with the resolution of the ones you had? I think there's still quite a bit of code duplication here but I think I'd be OK merging this one as is now and trying to polish the code in future PRs to prevent us from getting on a standstill with this one...

Cheers,
Filipe

**Comment by roberthbailey** [comment]
Sorry, I didn't see that my comments had been addressed. Looking now.

File: test/e2e/pods.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+		  "kind": "Pod",
+		  "apiVersion": "v1beta1",
+		  "id": "srv",
+		  "desiredState": {
+			"manifest": {
+			  "version": "v1beta1",
+			  "id": "srv",
+			  "containers": [{
+				"name": "srv",
+				"image": "kubernetes/serve_hostname",
```
Comment (roberthbailey) [inline test/e2e/pods.go]: Does this image have a specific tag we can reference?

File: test/e2e/pods.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+				"ports": [{
+				  "containerPort": 9376,
+				  "hostPort": 8080
```
Comment (roberthbailey) [inline test/e2e/pods.go]: Do we really need to use hostPort here?

**Comment by roberthbailey** [comment]
I agree that there is still a lot of code duplication that can be cleaned up, but I'll defer to you on whether that should block this getting in. I'm happy with the direction this is taking us (multiple specs per test and making the test groupings make logical sense).

**Comment by rrati** [comment]
@roberthbailey I addressed some of the issues you raised earlier but didn't want to differ from the originals too much code in this CR given the nature of the changes (filenames, moving content, etc).  I agree with all the comments you made.

**Comment by rrati** [comment]
@filbranden I'll rebase and push again

**Comment by rrati** [comment]
@filbranden Rebased branch pushed

**Comment by filbranden** [comment]
Thanks @rrati! Please keep the contributions coming!

**Comment by roberthbailey** [comment]
No worries. Looking forward to seeing them addressed in a future PR.

**Comment by satnam6502** [comment]
+1. Well, +N hopefully. Thanks.

File: test/e2e/pods.go (hunk 4 above)
Comment (satnam6502) [inline test/e2e/pods.go]: Yes it does. I think the current one is :1.1 and I encourage its use.
Note that serve_hostname (not tag, or :latest) is the oldest image and adds a newline character to the newline.
Note that serve_hostname:1.0 (based on ubuntu so big) or server_hostname:1.1 (based on scratch, so small) **do not** add a newline character after the hostname.

**Comment by rrati** [comment]
https://github.com/GoogleCloudPlatform/kubernetes/issues/3837

**Pull request merged**